### PR TITLE
rust: Mark functions in the module! macro as unsafe and add SAFETY comments

### LINKED ISSUE #629: Mark functions generated by the module! macro as unsafe
Most of them are unsafe despite not being marked as such. For example `fn __init` and `fn __exit` contain unsynchronized accesses to `static mut __MOD`. There are also several `extern "C"` functions which call `__init` and `__exit`. These need to be marked as unsafe too.

cc https://rust-for-linux.zulipchat.com/#narrow/stream/288089-General/topic/patch.20v3.20review/near/268298086

### PR DESCRIPTION
**Pull request by crapStone**
closes #629

Signed-off-by: Andreas Reindl <areindl@mailbox.org>

### CODE REVIEW

**Comment by bjorn3** [comment]
Thanks for the PR! You will have to sign off your comments. This is a requirement for submitting any patch to linux. See https://www.kernel.org/doc/html/latest/process/submitting-patches.html#sign-your-work-the-developer-s-certificate-of-origin for what this means and how to do this.

**Comment by crapStone** [comment]
@bjorn3 done

**Comment by bjorn3** [comment]
I believe you have to add that line to the individual commits. LGTM otherwise.

**Comment by crapStone** [comment]
Sorry, I thought the PR would get squashed. Did it myself now and added the `Signed-off-by` to the commit.

**Review by bjorn3 — APPROVED** [review]

**Comment by crapStone** [comment]
Should I ping someone to get their attention or are you just busy at the moment?

File: rust/macros/module.rs [hunk 1]
```diff
@@ -543,18 +544,28 @@ pub(crate) fn module(ts: TokenStream) -> TokenStream {
             }};
 
             // Loadable modules need to export the `{{init,cleanup}}_module` identifiers.
+            /// # Safety
+            /// This is a exported identifier that should only be called by the kernel at load time.
```
Comment (ojeda) [inline rust/macros/module.rs]: Initially I was going to comment that this safety section does not really explain much about the safety preconditions, but I guess it is actually a precondition, in the sense that it does not allow anybody else to call it, so it is fine.

In any case, nit on formatting -- the convention we use is:

```suggestion
            ///
            /// # Safety
            ///
            /// This is a exported identifier that should only be called by the kernel at load time.
```

File: rust/macros/module.rs [hunk 2]
```diff
@@ -576,23 +587,36 @@ pub(crate) fn module(ts: TokenStream) -> TokenStream {
...
 
+            /// # Safety
+            /// This function should only be use as a initcall pointer to called by the kernel at load time.
             #[cfg(not(MODULE))]
             #[doc(hidden)]
             #[no_mangle]
-            pub extern \"C\" fn __{name}_init() -> core::ffi::c_int {{
-                __init()
+            pub unsafe extern \"C\" fn __{name}_init() -> core::ffi::c_int {{
+                // SAFETY: This function will only be called by the kernel at load time.
```
Comment (ojeda) [inline rust/macros/module.rs:609]: Similarly, this comment is a bit atypical, but I guess it is fine on its own as long as it satisfies the safety precondition on `__init()`.

File: rust/macros/module.rs [hunk 3]
```diff
@@ -576,23 +587,36 @@ pub(crate) fn module(ts: TokenStream) -> TokenStream {
...
+            pub unsafe extern \"C\" fn __{name}_exit() {{
+                // SAFETY: This function will only be called by the kernel when the module gets unloaded.
+                unsafe {{
+                    __exit()
+                }}
             }}
 
-            fn __init() -> core::ffi::c_int {{
+            /// # Safety
+            /// This function should only be called by the kernel to ensure exclusive access to the local `static mut __MOD` variable.
```
Comment (ojeda) [inline rust/macros/module.rs]: Having said that, this one is not really called by the kernel (in the sense of the C side, like the others above), but from ourselves. Does it need to be clarified a bit? i.e. to distinguish it from the others above that are actually called from somewhere else.

In addition, normally there is no need to say why the precondition is required -- that is in the implementation (and its safety comments) that uses the precondition. In other words, the "to ensure..." part.

File: rust/macros/module.rs [hunk 4]
```diff
@@ -604,7 +628,10 @@ pub(crate) fn module(ts: TokenStream) -> TokenStream {
                 }}
             }}
 
-            fn __exit() {{
+            /// # Safety
+            /// This function should only be called by the kernel to ensure exclusive access to the local `static mut __MOD` variable.
+            unsafe fn __exit() {{
+                // SAFETY: This function will only be called by the kernel when the module gets unloaded which ensures, that we have exclusive access.
```
Comment (ojeda) [inline rust/macros/module.rs]: Nit: this line is a bit too long (134).

Likely others too -- please try to keep them under 100 lines, this is not enforced inside macros by `rustfmt`, but you may check with `checkpatch.pl`.

**Review by ojeda — COMMENTED** [review]
In case you didn't see it, a related PR is https://github.com/Rust-for-Linux/linux/pull/645.

**Comment by crapStone** [comment]
> In case you didn't see it, a related PR is #645.

Oh I didn't see it. It wasn't linked in the issue.

Should I close mine, because the other one was first?

**Comment by ojeda** [comment]
Not at all, please feel free to keep this one open. Ideally we would use both to get the best end result, which is why I mentioned it here, since you may want to see what the other person wrote in the safety comments, maybe discussing it with them if you don't agree with their approach, etc.

File: rust/macros/module.rs (hunk 1 above)
Comment (crapStone) [reply rust/macros/module.rs]: Done, fixed every `# Safety` comment

File: rust/macros/module.rs (hunk 2 above)
Comment (crapStone) [reply rust/macros/module.rs:609]: Done, explained it a bit more

File: rust/macros/module.rs (hunk 3 above)
Comment (crapStone) [reply rust/macros/module.rs]: Done, explained it a bit more

File: rust/macros/module.rs (hunk 4 above)
Comment (crapStone) [reply rust/macros/module.rs]: Done, tried to shorten every comment to 100 characters per line

**Comment by crapStone** [comment]
@ojeda can I get access to the Zulip chat?

**Comment by ojeda** [comment]
Could you please check if the sign up page works for you? (in the Zulip, please click on "Sign up") If that doesn't work, I am happy to send an invitation (to the email in your GitHub profile?)

**Pull request closed without merging**